babel/messages/frontend.py

```diff
@@ -852,6 +853,247 @@ def run(self):
+        ('sort-output', 's', 'generate sorted output'),
+        ('sort-by-file', 'F', 'sort output by file location'),
+    ]
+
+    as_args = 'input-files'
+
+    multiple_value_options = (
+        'compendium'
+    )
+
+    boolean_options = [
+        'compendium-overwrite',
+        'no-compendium-comment',
+        'update',
+        'backup',
+        'no-fuzzy-matching',
+        'no-location',
+        'no-wrap',
+        'sort-output',
+        'sort-by-file',
+    ]
+
+    def initialize_options(self):
+        self.input_files = None
+        self.compendium = None
+        self.compendium_overwrite = False
+        self.no_compendium_comment = False
+        self.update = False
+        self.output_file = None
+        self.backup = False
+        self.suffix = '~'
+        self.no_fuzzy_matching = False
+        self.no_location = False
+        self.width = None
+        self.no_wrap = False
+        self.sort_output = False
+        self.sort_by_file = False
+
+    def finalize_options(self):
+        if not self.input_files or len(self.input_files) != 2:
+            raise OptionError('must be two po files')
+        if not self.output_file and not self.update:
+            raise OptionError('you must specify the output file or update existing')
+
+        if self.no_wrap and self.width:
+            raise OptionError("'--no-wrap' and '--width' are mutually exclusive")
+        if not self.no_wrap and not self.width:
+            self.width = 76
+        elif self.width is not None:
+            self.width = int(self.width)
+
+    def _get_message_from_compendium(self, compendium):
+        for file_path in compendium:
+            with open(file_path, 'r') as pofile:
+                catalog = read_po(pofile)
+                for message in catalog:
+                    yield message, file_path
+
+    def run(self):
+        def_file, ref_file = self.input_files
+
+        with open(def_file, 'r') as pofile:
+            catalog = read_po(pofile)
+        with open(ref_file, 'r') as pofile:
+            ref_catalog = read_po(pofile)
+        catalog.update(
+            ref_catalog,
+            no_fuzzy_matching=self.no_fuzzy_matching
+        )
+
+        if self.compendium:
+            for message, compendium_path in self._get_message_from_compendium(self.compendium):
+                current = catalog[message.id]
+                if message.id in catalog and (not current.string or current.fuzzy or self.compendium_overwrite):
+                    if self.compendium_overwrite and not current.fuzzy and current.string:
+                        catalog.obsolete[message.id] = current.clone()
+
+                    current.string = message.string
+                    if current.fuzzy:
+                        current.flags.remove('fuzzy')
+
+                    if not self.no_compendium_comment:
+                        current.auto_comments.append(compendium_path)
```

Comment:
This could be unnested, and `get_messages_from_compendiums` could early-return if the passed argument is falsy.

```suggestion
        for message, compendium_path in self._get_message_from_compendium(self.compendium):
            current = catalog[message.id]
            if message.id in catalog and (not current.string or current.fuzzy or self.compendium_overwrite):
                if self.compendium_overwrite and not current.fuzzy and current.string:
                    catalog.obsolete[message.id] = current.clone()

                current.string = message.string
                if current.fuzzy:
                    current.flags.remove('fuzzy')

                if not self.no_compendium_comment:
                    current.auto_comments.append(compendium_path)
```